src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -1109,22 +1103,30 @@ trait Contexts { self: Analyzer =>
       def lookupImport(imp: ImportInfo, requireExplicit: Boolean) =
         importedAccessibleSymbol(imp, name, requireExplicit, record = true) filter qualifies
 
-      // Java: A single-type-import declaration d in a compilation unit c of package p
-      // that imports a type named n shadows, throughout c, the declarations of:
-      //
-      //  1) any top level type named n declared in another compilation unit of p
-      //
-      // A type-import-on-demand declaration never causes any other declaration to be shadowed.
-      //
-      // Scala: Bindings of different kinds have a precedence defined on them:
-      //
-      //  1) Definitions and declarations that are local, inherited, or made available by a
-      //     package clause in the same compilation unit where the definition occurs have
-      //     highest precedence.
-      //  2) Explicit imports have next highest precedence.
-      def depthOk(imp: ImportInfo) = (
-           imp.depth > symbolDepth
-        || (unit.isJava && imp.isExplicitImport(name) && imp.depth == symbolDepth)
+      /* Java: A single-type-import declaration d in a compilation unit c of package p
+       * that imports a type named n shadows, throughout c, the declarations of:
+       *
+       *  1) any top level type named n declared in another compilation unit of p
+       *
+       * A type-import-on-demand declaration never causes any other declaration to be shadowed.
+       *
+       * Scala: Bindings of different kinds have a precedence defined on them:
+       *
+       *  1) Definitions and declarations that are local, inherited, or made available by a
+       *     package clause in the same compilation unit where the definition occurs have
+       *     highest precedence.
+       *  2) Explicit imports have next highest precedence.
+       *  3) Wildcard imports have next highest precedence.
+       *  4) Definitions made available by a package clause not in the compilation unit where
+       *     the reference occurs have lowest precedence.
+       */
+      def depthOk(imp: ImportInfo) = imp.depth > symbolDepth || (
+           imp.isExplicitImport(name)
+        && imp.depth == symbolDepth
+        && (
+              unit.isJava
+           || (defSym != NoSymbol && isPackageOwnedInDifferentUnit(defSym))  // SI-2458
```

Comment:
I think we need to stay bug-compatible with the long standing behaviour under 2.12. I'd suggest predicating all the fixes in the PR on on `-Xsource:2.13`.

Reply:
"Copy that."